### DISCUSSION TITLE
[Proposal] Object initializers for factory methods

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#6602] [Language Ideas]
Specification: [proposals/factory-methods.md](https://github.com/dotnet/csharplang/blob/main/proposals/factory-methods.md)
Issue: https://github.com/dotnet/csharplang/issues/10292

-----

@ashmind commented on [Wed Jan 28 2015](https://github.com/dotnet/roslyn/issues/133)

#### Problem

This is another one of the issues when a change to logic significantly changes code structure.
Let's say I had the following:

``` csharp
var x = new X {
    A = a,
    B = b,
    C = { D = { d1, d2 } }
};
```

Now I decided all `X` should be created through an `XFactory` instead.
So I get:

``` csharp
var x = factory.CreateX();
x.A = a;
x.B = b;
x.C.D.Add(d1);
x.C.D.Add(d2);
```

One small logical change causes a total restructure of not-really-related code.
#### Potential solution

Provide a way to use object initializer with any existing object.
Example:

``` csharp
var x = factory.CreateX() with {
    A = a,
    B = b,
    C = { D = { d1, d2 } }
};
```

Keyword `with` here is provided to remove any syntax ambiguities.
##### Pros
1. More consistent syntax everywhere. Shorter syntax for setting a few properties when you need that.
##### Cons
1. New kind of expression-with-side-effects when used on an existing object.

---

@RichiCoder1 commented on [Wed Jan 28 2015](https://github.com/dotnet/roslyn/issues/133#issuecomment-71939498)

The con could be a big con, but it could be a pro if you expand the scope of the idea. Setting multiple properties on an object after it's been created is not an unheard of thing. 

---

@paulomorgado commented on [Fri Mar 06 2015](https://github.com/dotnet/roslyn/issues/133#issuecomment-77597563)

@ashmind,

Can you give an example of what syntax ambiguities does the use of `with` remove?

Can you elaborate on the **cons** part?

Are you aware that this:

```
var x = new X {
    A = a,
    B = b,
    C = { D = { d1, d2 } }
};
```

is translated into this?

```
X tmp = new X();
tmp.A = a;
tmp.B = b;
tmp.C.D.Add(d1);
tmp.C.D.Add(d1);
X x = tmp;
```

The use cases for a block assignment are many. One is the assign properties of return values of methods. But I could also do something like this:

```
var x = new X { A = 1, B = 2 };
var y = x { A = a, B = b, C = { D = { d1, d2 } } };
```

I just think you got the motivation all wrong.

---

@ashmind commented on [Fri Mar 06 2015](https://github.com/dotnet/roslyn/issues/133#issuecomment-77633733)

> Can you give an example of what syntax ambiguities does the use of `with` remove?

I'm not sure if there are any in the spec at the moment, but allowing `x { A = 3 }` for arbitrary `x` seems confusing and may block some other features from being introduced in the future.

> Can you elaborate on the cons part?

E.g.

``` csharp
var x = new X();
var y = GetSomething(1, 2, x with { A = 5 }, 4, 5);
// x here has A = 5, which is not immediately obvious
```

I suppose it can be avoided if initializers are limited to method call results only.

> Are you aware that this / is translated into this?

Yes, however I don't want to write it myself for factory methods, given that translator is already in the language.

> I just think you got the motivation all wrong.

Can you elaborate? Is your motivation mostly about allowing modification of existing objects?

---

@paulomorgado commented on [Sun Mar 08 2015](https://github.com/dotnet/roslyn/issues/133#issuecomment-77773457)

@ashmind, I still find the whole thing lacks consistency. You talk about side effects when there aren't any, in comparison.

Sure shorthand would be nice. But that's just it. There are no other pros or cons involved.

---

@amironov commented on [Wed May 06 2015](https://github.com/dotnet/roslyn/issues/133#issuecomment-99408183)

The proposal also solves the problem of [object initializers in `using (IDisposable)` statement](http://ayende.com/blog/3810/avoid-object-initializers-the-using-statement).

Syntax can be similar to the existing (compare `x = {...}` and `C = {...}`):

``` C#
using (var x = new X()) {
    x = {
        A = 1,
        B = 2,
        C = { D = { 3, 4 } },
    };
}
```

---

@Licshee commented on [Wed May 25 2016](https://github.com/dotnet/roslyn/issues/133#issuecomment-221489862)

IMHO ~~we this problem should be split into two, or~~ at least there are quite a few questions we are discussing:
1) removing "with" keyword should not lead to ambiguity, but only if applied immediately after a method call, it's a visual/readability thing;
2) allow initializer syntax to show up immediately after a variable or field/property doesn't look like a good idea, mostly b/c when we use initializer syntax, we are usually in a situation where the result should be available as soon as possible, like:

```
cmd.Parameters.Add(cmd.CreateParameter() { ParameterName = "@name", Value = value });
```

3) but there are cases where applying initializer syntax to a property can be useful, or even a variable, i.e.:

```
cmd.Parameters {
    cmd.CreateParameter() { ParameterName = "@key", Value = key },
    cmd.CreateParameter() { ParameterName = "@value", Value = value }, 
};
```

and this lacks some sort of readability;

So far the original idea from @ashmind is the best solution for the questions listed above.

---

@MadsTorgersen commented on [Mon Aug 15 2016](https://github.com/dotnet/roslyn/issues/133#issuecomment-239888215)

It's desirable to allow the same initializers on factories as constructor calls. We need to think this through in conjunction with `with` expressions, and with object initializers for immutable objects.

---

@Spongman commented on [Mon Jan 30 2017](https://github.com/dotnet/roslyn/issues/133#issuecomment-276190940)

i agree that the 'with' keyword seems unnecessary. i'd like to see this simple extension of 'object initializers' to arbitrary expressions:
1. new Foo { A = 1 }
2. new Foo () { A = 1 }
3. foo { A = 1}
3. (GetBar().Foos[3]) { A = 1}

**Comment by yaakov-h** [comment]
"Generating" a null reference exception is not a null checking behaviour. It's the complete absence of one.

I think `with?` would be a bit more C#-y if we were to go that route.

**Comment by bahusoid** [comment]
While it's not implemented you can use this [approach ](https://stackoverflow.com/a/55086075/9667775) to use object initializer with any existing object.

You should create a simple object wrapper:

```C#
public struct ObjectIniter<TObject>
{
    public ObjectIniter(TObject obj)
    {
        Obj = obj;
    }

    public TObject Obj { get; }
}
```
And now you can use it like this to initialize your objects:

```C#
new ObjectIniter<MyClass>(existingInstance)
{
    Obj =
    {
        //Object initializer of MyClass:
        Property1 = value1,
        Property2 = value2,
        //...
    }
};
```

**Comment by mikernet** [comment]
@ghost @bahusoid Those are some clever tricks haha, well done... +1 for creativity :)

**Comment by julealgon** [comment]
@gafter / @MadsTorgersen any news on this proposal? With record types now released, I assume some of the discussions would have continued here.

**Comment by miyu** [comment]
Are these two cases identical?

```csharp
// case 1
var x = MyFactory() with { X = 123 };
// case 2
var temp = MyFactory();
var x = temp with { X = 123 }; // is this legal? is temp == x?
```

If the two are identical, then is `ReferenceEquals(x, temp)` true?

* Yes (`x == temp`): then this breaks my mental-model of `with` as a struct copy & some property/field sets & code becomes confusing to read; it'll be hard to know whether the left-operand of `with` is mutated or not.  

    * Does `Console.WriteLine(someVariable with { X = 1 })` mutate someVariable? It depends on if it's a struct or class!?

* No (`x != temp`): then objects will need to implement a cloning operation of some sorts, right?

Maybe we need a different keyword or mechanism for marking factory methods as `with`-friendly?

**Comment by FaustVX** [comment]
@miyu 
It would be weird to have 2 different meaning.
Currently with `record` or `struct` we have that:
```C#
record R(string S);
var r0 = new R("1");
var r1 = r0 with { S = "2" }
// r1 != r2
```

**Comment by miyu** [comment]
A few additional complexities in the case where `with` clones:

* Large performance hit vs simply assigning fields for little benefit?
* ~~Records are nonpolymorphic, whereas return values of factories can be.~~ How is the clone dispatched?
* Factory objects can't necessarily be cloned (e.g. what if an instance owns a socket?); we'd probably want ownership-transfer semantics (like C++ move) but implementing that is quite cumbersome. In any case, that's still different vs what `with` currently does.
* Cloning isn't necessarily obvious. This seems benign: `File.Open(...) with { Position = 123, WriteTimeout = 100 }`?

Personally I'd prefer this not involve a clone & that it use a different operator than `with`. I like `init`, since we already have `init-only setters` and presumably we want to allow setting these within factory method initializers.

```
var panel = GUIFactory.CreatePanel() init {
  Text = "Hi"
}
// equivalent
var panel = GUIFactory.CreatePanel();
panel.Text = "Hi";
```

Some examples:
* `var panel = GUIFactory.CreatePanel() . { Text = "" }`
* `var panel = GUIFactory.CreatePanel() .. { Text = "" }`
* `var panel = GUIFactory.CreatePanel() << { Text = "" }`
* `var panel = GUIFactory.CreatePanel() set { Text = "" }`
* `var panel = GUIFactory.CreatePanel() init { Text = "" }`
* `var panel = GUIFactory.CreatePanel() assign { Text = "" }`

**Reply by jnm2** [reply]
C# records are polymorphic ([link](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#with-expressions-in-derived-records)).

**Reply by mikernet** [reply]
Allowing init props to be set after a factory method call would break init props. You would be able to tack on `init { ... }` after a call to any method that returns any object and set init props on that object well after it has been created.

**Reply by Haukinger** [reply]
Not if you annotate the factory method as such.

I'd love to see `someFactory.CreateProduct("param") { Prop = 3 }` being an extension of `new Something("param") { Prop = 3 }`.

**Reply by Spongman** [reply]
you missed one (the best one IMO):

```
var panel = GUIFactory.CreatePanel() { Text = "" }
```

**Comment by jnm2** [comment]
I've moved @jaredpar's original proposal to this repo which was [discussed in a 2020 LDM](https://github.com/dotnet/csharplang/blob/main/meetings/2020/LDM-2020-04-20.md):
- Specification: [proposals/factory-methods.md](https://github.com/dotnet/csharplang/blob/main/proposals/factory-methods.md)
- Issue: https://github.com/dotnet/csharplang/issues/10292

**Reply by timcassell** [reply]
I can't tell if the proposal allows for async factories. Please consider it.

Also, if factories are "treated like constructors", why disallow this?

```cs
    [Factory]
    public Student Create4()
    {
        var s = new Student();
        return s;
    }
```

I would find it useful to be able to create an object (or retrieve from a pool like @ghord's comment), call some method, then assign properties before returning the initialized object. All of that is supported in constructors.

**Reply by AbsoluteMoose** [reply]
The factory method restrictions in the current proposal seem very limiting. I often use a factory method rather than a constructor when an object requires additional setup that cannot be performed in a constructor, such as creating a type from a third-party library that requires using methods to initialize different aspects of the object, performing some setup that needs to be done after all constructors have run on a derived class, or instantiating multi-part objects or object trees. 

An example of a pattern that would greatly benefit from factory initializers but would be excluded under these restrictions is Godot Engine's PackedScene.Instantiate(). That method returns a new object with parameters defined by the contents of a scene file, potentially including child objects essential to the functionality of the main object. It requires interop with Godot's C++ code in a way that makes it impossible to directly return a new object, but from an end user perspective it's an entirely new object.

**Comment by ghord** [comment]
I think that restrictions on factory methods are too limiting in this proposal. I was hoping that introducing Factory methods would allow for pooling of objects. Please consider following code:

```c#
[Factory]
MyElement Create()
{
    var element = _myCustomPool.Get<MyElement>(); // could be for example some reusable pool of objects
    return element;
}
```

Reducing allocations as much as possible on object construction could help in initiative such as declarative UI construction.

I feel that not making factories work with such pattern would would be a big miss.

**Comment by vladd** [comment]
For me, the primary usecase of factory methods is working around the impossibility to use `await` in constructor body.
Please consider async factories as well.

**Comment by vladd** [comment]
Can a factory for a class be defined in a different assembly?

If yes, is this going to work with the current proposal?

```csharp
// Assembly A
class X
{
    public int V { get; init; }
}

// Assembly B
X x = new X { V = 1 };
//x.V = 2; // doesn't compile
Mutate(x, 2); // works?

void Mutate(X x, int newv) => _ = Id(x) { V = newv };

[Factory]
X Id(X x) => x;
```

**Comment by azchohfi** [comment]
I prototyped the return-expression restriction from the championed proposal (#10292 / `factory-methods.md`) in a local Roslyn build and applied `[Factory]` mechanically across [Microsoft.UI.Reactor](https://github.com/microsoft/microsoft-ui-reactor) — an MVU UI framework for WinUI, ~200 factory methods — to see what it catches. Two bits of feedback, one of which looks like a genuine gap rather than a cost.

**Factory-ness can't cross a delegate boundary.**

The restriction allows a `new` expression, a call to another `[Factory]` member, a struct value, or a ternary/coalescing expression whose branches qualify. A delegate invocation is none of those, and there's nowhere to put the attribute — the callee is supplied by the caller at runtime:

```csharp
[Factory]
public static Element When(bool condition, Func<Element> then) =>
    condition ? then() : Empty();     // `then()` can never satisfy the restriction
```

(that's [Reactor's actual `When`](https://github.com/microsoft/microsoft-ui-reactor/blob/469f5ef4d0c3f8b8a07590bfa75054e42299b487/src/Reactor/Elements/Dsl.cs#L1259), modulo the attribute)

This isn't fixable by annotating more members: `then` is a parameter. Any factory that defers construction to a caller-supplied function is permanently excluded. The `Func<>` is usually there for laziness — don't build the subtree unless you need it — so it's not an odd shape.

It generalises well past UI: DI container registrations (`services.AddSingleton(sp => new Foo(...))`), `Lazy<T>` factories, object-pool creation callbacks, builder APIs taking `Func<T>`. All of those are "make me a fresh thing" functions, which is exactly the population `[Factory]` is about, and none of them can participate.

**One shape that would close it** — put the promise on the delegate *type* and check it at conversion:

```csharp
[Factory] public delegate Element ElementFactory();

[Factory]
public static Element When(bool condition, ElementFactory then) =>
    condition ? then() : Empty();
```

Inside `When`, `then()` qualifies because the delegate type carries the promise. At each conversion into `ElementFactory`, the compiler checks the lambda body or method group against the same restriction — `() => Text("hi")` passes if `Text` is `[Factory]`; `() => _cached` is rejected at the call site, which is where the mistake actually is.

It has to be the type rather than the parameter, I think: a parameter-only annotation is lost the moment the delegate is stored in a plain `Func<T>` field, whereas a type carries the contract across assignment. Downside is the obvious one — `Func<T>` and `ElementFactory` stop being interchangeable, so existing helpers returning `Func<T>` don't flow in.

**Separately, on the cost of the restriction** (context rather than an objection): applying `[Factory]` to Reactor and iterating to a fixpoint took **~876 annotations**, not the ~200 I expected. Fluent modifiers are `=> ModifyLayout(el, …)`, so annotating a factory pulls in its callee, and that callee's callee — and the chain terminates in `private` helpers. Worth knowing that the attribute propagates into implementation details that aren't part of any public contract; an analyzer that applies the closure automatically would probably be necessary in practice.

For what it's worth, the restriction did its job there: it flagged six methods that genuinely hand out a shared cached instance, which is exactly the bug the rule exists to prevent.

Is the delegate case something you'd want to handle, or is "factory-ness stops at a delegate boundary" an accepted limitation?
